impress-org/givewp-next-gen
# Refactor: register and load form template scripts in cascading order

## Description



This is a big architecture update on how form template related scripts are loaded.  The new loading order looks like this:

**1. Registrar scripts**

This first script loads to assign `window.givewp.form.templates = Object.freeze(defaultFormTemplates)`.  
Our `defaultFormTemplates` is the starting point for add-ons and form designs to interact with.  We make sure to freeze the initial object so that scripts downstream can only modify templates directly to preserve data integrity as much as possible.

**2. Active Form Design**

The active form design (if supplying a JS file) will load _after_ our registrar scripts, so it will have direct access to givewp using the window data:

```
// Get the current version of a template;
const Goal = window.givewp.form.templates.layouts.goal // FC<GoalProps> --> this is being retrieved from core

// Modify the template
window.givewp.form.templates.layouts.goal = MyNewGoal
```

**4. Form Application.**

When the donation form application loads, it simply uses `window.givewp.form.templates` to retrieve the final version of the formTemplates (which is the result of merging core & the active templates - later this will include add-ons).  Throughout the application we are now free to use the templates from the window and grab the value of the template we want to use.

```
const formTemplates = window.givewp.form.templates;

const HeaderTemplate = withTemplateWrapper(formTemplates.layouts.header);
const HeaderTitleTemplate = withTemplateWrapper(formTemplates.layouts.headerTitle);
const HeaderDescriptionTemplate = withTemplateWrapper(formTemplates.layouts.headerDescription);
const GoalTemplate = withTemplateWrapper(formTemplates.layouts.goal);

// render
<HeaderTemplate />
```


**Template Wrapper**

Going down this path allowed me to simplify our rendering process.  Instead of having specific functions for retrieving and rendering specific types of templates, we will now simply grab the templates directly from the active form design and use these template wrappers below.  These will determine how to wrap the template from the template value itself.  The main advantage of this direction is everything becomes self-documenting because the value of the template you are retrieving is typed and used as the generic.

- `withTemplateWrapper`
- `useTemplateWrapper`
- `<TemplateWrapper />`

```
const FieldTemplate = useTemplateWrapper<FieldProps>(formTemplates.fields[node.type]); // React.FC<FieldProps>
const GatewayFieldTemplate = withTemplateWrapper(formTemplates.fields.gateways); // React.FC<GatewayFieldProps>
```


## Affects



There is no visual change but this does affect the whole form rendering process.

## Visuals



There are no visual changes.

## Testing Instructions



Build scripts and make sure the form rendering works and preview work as expected.  There should be no visual differences.

## Pre-review Checklist
